wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,270 @@
+
+        -- If this is a hard coin selection error we cannot recover, stop
+        -- retrying. If we get a 'Tx' as output, stop retrying immediately.
+        shouldRetry :: RetryStatus -> Either PaymentError Tx -> IO Bool
+        shouldRetry _ (Right _)                             = return False
+        shouldRetry _ (Left (PaymentNewTransactionError _)) = return False
+        shouldRetry _ _                                     = return True
+
+-- | Creates a new 'TxAux' without submitting it to the network.
+newTransaction :: ActiveWallet
+               -> IO Address
+               -- ^ A computation to generate change 'Address'es.
+               -> (Address -> Either CoinSelHardErr SafeSigner)
+               -- ^ A function to sign each Address
+               -> CoinSelectionOptions
+               -- ^ The options describing how to tune the coin selection.
+               -> HdAccountId
+               -- ^ The source @(root, account) from where the payment was
+               -- originated
+               -> NonEmpty (Address, Coin)
+               -- ^ The payees
+               -> IO (Either NewTransactionError TxAux)
+newTransaction ActiveWallet{..} genChangeAddr signAddress options accountId payees = do
+    snapshot <- getWalletSnapshot walletPassive
+    let toTxOuts = fmap (\(a,c) -> TxOutAux (TxOut a c))
+    let mkTx = mkStdTx walletProtocolMagic signAddress
+    -- | NOTE(adn) This number was computed out of the work Matt Noonan did
+    -- on the size estimation. See [CBR-318].
+    let maxInputs = 350
+    let availableUtxo = accountAvailableUtxo snapshot accountId
+    res <- flip runReaderT payees . buildPayment $
+           CoinSelection.random options
+                                (liftIO genChangeAddr)
+                                mkTx
+                                maxInputs
+                                (toTxOuts payees)
+                                availableUtxo
+    case res of
+         Left err -> return . Left . CoinSelectionFailed $ err
+         Right t  -> return . Right $ t
+
+-- | Special monad used to process the payments, which randomness is derived
+-- from a fixed seed obtained from hashing the payees. This guarantees that
+-- when we estimate the fees and later create a transaction, the coin selection
+-- will always yield the same value, making the process externally-predicatable.
+newtype PayMonad a = PayMonad {
+      buildPayment :: ReaderT (NonEmpty (Address, Coin)) IO a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadReader (NonEmpty (Address, Coin)))
+
+-- | \"Invalid\" 'MonadRandom' instance for 'PayMonad' which generates
+-- randomness using the hash of 'NonEmpty (Address, Coin)' as fixed seed.
+instance MonadRandom PayMonad where
+    getRandomBytes _ =
+        ask >>= return . ByteArray.convert
+                       . encodeUtf8 @Text @ByteString
+                       . sformat build
+                       . hash
```

Comment:
This is indeed bollocks, as the initial seed is correctly computed, but successive calls to `getRandomBytes` yields exactly the same result, so that's no randomness at all here. We should probably use the hash plus something like an incremental counter tucked into an `IORef` to ensure we get deterministic yet random results.